// zod-geojson/lib/geometry/geometry_collection.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeoJSON3DGeometryCollectionSchema = exports.GeoJSON2DGeometryCollectionSchema = exports.GeoJSONGeometryCollectionSchema = exports.GeoJSONGeometryCollectionGenericSchema = void 0;
const z = __importStar(require("zod"));
const base_1 = require("../base.js");
const geometry_1 = require("./geometry.js");
const position_1 = require("./position.js");
const type_1 = require("./type.js");
const bbox_1 = require("./validation/bbox.js");
const dimension_1 = require("./validation/dimension.js");
const GeoJSONGeometryCollectionGenericSchema = (positionSchema) => z
    .looseObject({
    ...(0, base_1.GeoJSONBaseSchema)(positionSchema).shape,
    type: z.literal(type_1.GeoJSONGeometryTypeSchema.enum.GeometryCollection),
    coordinates: z.never({ error: "GeoJSON geometry collection cannot have a 'coordinates' key" }).optional(),
    geometry: z.never({ error: "GeoJSON geometry collection cannot have a 'geometry' key" }).optional(),
    properties: z.never({ error: "GeoJSON geometry collection cannot have a 'properties' key" }).optional(),
    features: z.never({ error: "GeoJSON geometry collection cannot have a 'features' key" }).optional(),
    // > The value of "geometries" is an array. Each element of this array is a
    //   GeoJSON Geometry object. It is possible for this array to be empty. (RFC 7946, section 3.1.8)
    // > To maximize interoperability, implementations SHOULD avoid nested
    //    GeometryCollections. (RFC 7946, section 3.1.8)
    get geometries() {
        return z.array((0, geometry_1.GeoJSONGeometryGenericSchema)(positionSchema));
    },
})
    .check((ctx) => {
    // Skip remaining checks if geometries array is empty
    if (!ctx.value.geometries.length) {
        return;
    }
    if (!(0, dimension_1.validDimensionsForCollection)(ctx.value)) {
        ctx.issues.push((0, dimension_1.getInvalidGeometryCollectionDimensionIssue)(ctx));
    }
    if (!(0, bbox_1.validBBoxForCollection)(ctx.value)) {
        ctx.issues.push((0, bbox_1.getInvalidBBoxIssue)(ctx));
    }
});
exports.GeoJSONGeometryCollectionGenericSchema = GeoJSONGeometryCollectionGenericSchema;
exports.GeoJSONGeometryCollectionSchema = (0, exports.GeoJSONGeometryCollectionGenericSchema)(position_1.GeoJSONPositionSchema);
exports.GeoJSON2DGeometryCollectionSchema = (0, exports.GeoJSONGeometryCollectionGenericSchema)(position_1.GeoJSON2DPositionSchema);
exports.GeoJSON3DGeometryCollectionSchema = (0, exports.GeoJSONGeometryCollectionGenericSchema)(position_1.GeoJSON3DPositionSchema);
