// zod-geojson/lib/geometry/geometry_collection.mjs
import * as z from "zod";
import { GeoJSONBaseSchema } from "../base.mjs";
import { GeoJSONGeometryGenericSchema } from "./geometry.mjs";
import { GeoJSON2DPositionSchema, GeoJSON3DPositionSchema, GeoJSONPositionSchema, } from "./position.mjs";
import { GeoJSONGeometryTypeSchema } from "./type.mjs";
import { getInvalidBBoxIssue, validBBoxForCollection } from "./validation/bbox.mjs";
import { getInvalidGeometryCollectionDimensionIssue, validDimensionsForCollection } from "./validation/dimension.mjs";
export const GeoJSONGeometryCollectionGenericSchema = (positionSchema) => z
    .looseObject({
    ...GeoJSONBaseSchema(positionSchema).shape,
    type: z.literal(GeoJSONGeometryTypeSchema.enum.GeometryCollection),
    coordinates: z.never({ error: "GeoJSON geometry collection cannot have a 'coordinates' key" }).optional(),
    geometry: z.never({ error: "GeoJSON geometry collection cannot have a 'geometry' key" }).optional(),
    properties: z.never({ error: "GeoJSON geometry collection cannot have a 'properties' key" }).optional(),
    features: z.never({ error: "GeoJSON geometry collection cannot have a 'features' key" }).optional(),
    // > The value of "geometries" is an array. Each element of this array is a
    //   GeoJSON Geometry object. It is possible for this array to be empty. (RFC 7946, section 3.1.8)
    // > To maximize interoperability, implementations SHOULD avoid nested
    //    GeometryCollections. (RFC 7946, section 3.1.8)
    get geometries() {
        return z.array(GeoJSONGeometryGenericSchema(positionSchema));
    },
})
    .check((ctx) => {
    // Skip remaining checks if geometries array is empty
    if (!ctx.value.geometries.length) {
        return;
    }
    if (!validDimensionsForCollection(ctx.value)) {
        ctx.issues.push(getInvalidGeometryCollectionDimensionIssue(ctx));
    }
    if (!validBBoxForCollection(ctx.value)) {
        ctx.issues.push(getInvalidBBoxIssue(ctx));
    }
});
export const GeoJSONGeometryCollectionSchema = GeoJSONGeometryCollectionGenericSchema(GeoJSONPositionSchema);
export const GeoJSON2DGeometryCollectionSchema = GeoJSONGeometryCollectionGenericSchema(GeoJSON2DPositionSchema);
export const GeoJSON3DGeometryCollectionSchema = GeoJSONGeometryCollectionGenericSchema(GeoJSON3DPositionSchema);
