// zod-geojson/lib/geometry/line_string.mjs
import * as z from "zod";
import { GeoJSONGeometryBaseSchema } from "./helper/base.mjs";
import { GeoJSON2DPositionSchema, GeoJSON3DPositionSchema, GeoJSONPositionSchema, } from "./position.mjs";
import { GeoJSONGeometryTypeSchema } from "./type.mjs";
import { getInvalidBBoxIssue, validBBoxForPositionList } from "./validation/bbox.mjs";
import { getInvalidDimensionIssue, validDimensionsForPositionList } from "./validation/dimension.mjs";
export const GeoJSONLineStringGenericSchema = (positionSchema) => z
    .looseObject({
    ...GeoJSONGeometryBaseSchema(positionSchema).shape,
    type: z.literal(GeoJSONGeometryTypeSchema.enum.LineString),
    // > For type "LineString", the "coordinates" member is an array of two or
    //   more positions. (RFC 7946, section 3.1.4)
    coordinates: positionSchema.array().min(2),
})
    .check((ctx) => {
    // Skip remaining checks if coordinates empty
    if (!ctx.value.coordinates.length) {
        return;
    }
    if (!validDimensionsForPositionList(ctx.value)) {
        ctx.issues.push(getInvalidDimensionIssue(ctx));
        return;
    }
    if (!validBBoxForPositionList(ctx.value)) {
        ctx.issues.push(getInvalidBBoxIssue(ctx));
        return;
    }
});
export const GeoJSONLineStringSchema = GeoJSONLineStringGenericSchema(GeoJSONPositionSchema);
export const GeoJSON2DLineStringSchema = GeoJSONLineStringGenericSchema(GeoJSON2DPositionSchema);
export const GeoJSON3DLineStringSchema = GeoJSONLineStringGenericSchema(GeoJSON3DPositionSchema);
