// zod-geojson/lib/geometry/multi_polygon.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeoJSON3DMultiPolygonSchema = exports.GeoJSON2DMultiPolygonSchema = exports.GeoJSONMultiPolygonSchema = exports.GeoJSONMultiPolygonGenericSchema = void 0;
const z = __importStar(require("zod"));
const base_1 = require("./helper/base.js");
const polygon_1 = require("./polygon.js");
const position_1 = require("./position.js");
const type_1 = require("./type.js");
const bbox_1 = require("./validation/bbox.js");
const dimension_1 = require("./validation/dimension.js");
const linear_ring_1 = require("./validation/linear_ring.js");
const GeoJSONMultiPolygonGenericSchema = (positionSchema) => z
    .looseObject({
    ...(0, base_1.GeoJSONGeometryBaseSchema)(positionSchema).shape,
    type: z.literal(type_1.GeoJSONGeometryTypeSchema.enum.MultiPolygon),
    // We allow an empty coordinates array:
    // > GeoJSON processors MAY interpret Geometry objects with empty "coordinates"
    //   arrays as null objects. (RFC 7946, section 3.1)
    coordinates: (0, polygon_1.GeoJSONPolygonGenericSchema)(positionSchema).shape.coordinates.array(),
})
    .check((ctx) => {
    // Skip remaining checks if coordinates array is empty
    if (!ctx.value.coordinates.length) {
        return;
    }
    if (!(0, dimension_1.validDimensionsForPositionGridList)(ctx.value)) {
        ctx.issues.push((0, dimension_1.getInvalidDimensionIssue)(ctx));
        return;
    }
    if (!(0, linear_ring_1.validMultiPolygonLinearRings)(ctx.value)) {
        ctx.issues.push((0, linear_ring_1.getInvalidMultiPolygonLinearRingIssue)(ctx));
        return;
    }
    if (!(0, bbox_1.validBBoxForPositionGridList)(ctx.value)) {
        ctx.issues.push((0, bbox_1.getInvalidBBoxIssue)(ctx));
        return;
    }
});
exports.GeoJSONMultiPolygonGenericSchema = GeoJSONMultiPolygonGenericSchema;
exports.GeoJSONMultiPolygonSchema = (0, exports.GeoJSONMultiPolygonGenericSchema)(position_1.GeoJSONPositionSchema);
exports.GeoJSON2DMultiPolygonSchema = (0, exports.GeoJSONMultiPolygonGenericSchema)(position_1.GeoJSON2DPositionSchema);
exports.GeoJSON3DMultiPolygonSchema = (0, exports.GeoJSONMultiPolygonGenericSchema)(position_1.GeoJSON3DPositionSchema);
