// zod-geojson/lib/geometry/multi_line_string.mjs
import * as z from "zod";
import { GeoJSONGeometryBaseSchema } from "./helper/base.mjs";
import { GeoJSONLineStringGenericSchema } from "./line_string.mjs";
import { GeoJSON2DPositionSchema, GeoJSON3DPositionSchema, GeoJSONPositionSchema, } from "./position.mjs";
import { GeoJSONGeometryTypeSchema } from "./type.mjs";
import { getInvalidBBoxIssue, validBBoxForPositionGrid } from "./validation/bbox.mjs";
import { getInvalidDimensionIssue, validDimensionsForPositionGrid } from "./validation/dimension.mjs";
export const GeoJSONMultiLineStringGenericSchema = (positionSchema) => z
    .looseObject({
    ...GeoJSONGeometryBaseSchema(positionSchema).shape,
    type: z.literal(GeoJSONGeometryTypeSchema.enum.MultiLineString),
    // We allow an empty coordinates array
    // > GeoJSON processors MAY interpret Geometry objects with empty "coordinates"
    //   arrays as null objects. (RFC 7946, section 3.1)
    coordinates: GeoJSONLineStringGenericSchema(positionSchema).shape.coordinates.array(),
})
    .check((ctx) => {
    // Skip remaining checks if coordinates array is empty
    if (!ctx.value.coordinates.length) {
        return;
    }
    if (!validDimensionsForPositionGrid(ctx.value)) {
        ctx.issues.push(getInvalidDimensionIssue(ctx));
        return;
    }
    if (!validBBoxForPositionGrid(ctx.value)) {
        ctx.issues.push(getInvalidBBoxIssue(ctx));
        return;
    }
});
export const GeoJSONMultiLineStringSchema = GeoJSONMultiLineStringGenericSchema(GeoJSONPositionSchema);
export const GeoJSON2DMultiLineStringSchema = GeoJSONMultiLineStringGenericSchema(GeoJSON2DPositionSchema);
export const GeoJSON3DMultiLineStringSchema = GeoJSONMultiLineStringGenericSchema(GeoJSON3DPositionSchema);
