// zod-geojson/lib/geometry/multi_point.mjs
import * as z from "zod";
import { GeoJSONGeometryBaseSchema } from "./helper/base.mjs";
import { GeoJSON2DPositionSchema, GeoJSON3DPositionSchema, GeoJSONPositionSchema, } from "./position.mjs";
import { GeoJSONGeometryTypeSchema } from "./type.mjs";
import { getInvalidBBoxIssue, validBBoxForPositionList } from "./validation/bbox.mjs";
import { getInvalidDimensionIssue, validDimensionsForPositionList } from "./validation/dimension.mjs";
export const GeoJSONMultiPointGenericSchema = (positionSchema) => z
    .looseObject({
    ...GeoJSONGeometryBaseSchema(positionSchema).shape,
    type: z.literal(GeoJSONGeometryTypeSchema.enum.MultiPoint),
    // We allow an empty coordinates array
    // > GeoJSON processors MAY interpret Geometry objects with empty "coordinates"
    //   arrays as null objects. (RFC 7946, section 3.1)
    coordinates: z.array(positionSchema),
})
    .check((ctx) => {
    // Skip remaining checks if coordinates empty
    if (!ctx.value.coordinates.length) {
        return;
    }
    if (!validDimensionsForPositionList(ctx.value)) {
        ctx.issues.push(getInvalidDimensionIssue(ctx));
        return;
    }
    if (!validBBoxForPositionList(ctx.value)) {
        ctx.issues.push(getInvalidBBoxIssue(ctx));
        return;
    }
});
export const GeoJSONMultiPointSchema = GeoJSONMultiPointGenericSchema(GeoJSONPositionSchema);
export const GeoJSON2DMultiPointSchema = GeoJSONMultiPointGenericSchema(GeoJSON2DPositionSchema);
export const GeoJSON3DMultiPointSchema = GeoJSONMultiPointGenericSchema(GeoJSON3DPositionSchema);
