// zod-geojson/lib/geometry/multi_polygon.mjs
import * as z from "zod";
import { GeoJSONGeometryBaseSchema } from "./helper/base.mjs";
import { GeoJSONPolygonGenericSchema } from "./polygon.mjs";
import { GeoJSON2DPositionSchema, GeoJSON3DPositionSchema, GeoJSONPositionSchema, } from "./position.mjs";
import { GeoJSONGeometryTypeSchema } from "./type.mjs";
import { getInvalidBBoxIssue, validBBoxForPositionGridList } from "./validation/bbox.mjs";
import { getInvalidDimensionIssue, validDimensionsForPositionGridList } from "./validation/dimension.mjs";
import { getInvalidMultiPolygonLinearRingIssue, validMultiPolygonLinearRings } from "./validation/linear_ring.mjs";
export const GeoJSONMultiPolygonGenericSchema = (positionSchema) => z
    .looseObject({
    ...GeoJSONGeometryBaseSchema(positionSchema).shape,
    type: z.literal(GeoJSONGeometryTypeSchema.enum.MultiPolygon),
    // We allow an empty coordinates array:
    // > GeoJSON processors MAY interpret Geometry objects with empty "coordinates"
    //   arrays as null objects. (RFC 7946, section 3.1)
    coordinates: GeoJSONPolygonGenericSchema(positionSchema).shape.coordinates.array(),
})
    .check((ctx) => {
    // Skip remaining checks if coordinates array is empty
    if (!ctx.value.coordinates.length) {
        return;
    }
    if (!validDimensionsForPositionGridList(ctx.value)) {
        ctx.issues.push(getInvalidDimensionIssue(ctx));
        return;
    }
    if (!validMultiPolygonLinearRings(ctx.value)) {
        ctx.issues.push(getInvalidMultiPolygonLinearRingIssue(ctx));
        return;
    }
    if (!validBBoxForPositionGridList(ctx.value)) {
        ctx.issues.push(getInvalidBBoxIssue(ctx));
        return;
    }
});
export const GeoJSONMultiPolygonSchema = GeoJSONMultiPolygonGenericSchema(GeoJSONPositionSchema);
export const GeoJSON2DMultiPolygonSchema = GeoJSONMultiPolygonGenericSchema(GeoJSON2DPositionSchema);
export const GeoJSON3DMultiPolygonSchema = GeoJSONMultiPolygonGenericSchema(GeoJSON3DPositionSchema);
