// zod-geojson/lib/geometry/polygon.mjs
import * as z from "zod";
import { GeoJSONGeometryBaseSchema } from "./helper/base.mjs";
import { GeoJSON2DPositionSchema, GeoJSON3DPositionSchema, GeoJSONPositionSchema, } from "./position.mjs";
import { GeoJSONGeometryTypeSchema } from "./type.mjs";
import { getInvalidBBoxIssue, validBBoxForPositionGrid } from "./validation/bbox.mjs";
import { getInvalidDimensionIssue, validDimensionsForPositionGrid } from "./validation/dimension.mjs";
import { getInvalidPolygonLinearRingIssue, validPolygonRings } from "./validation/linear_ring.mjs";
export const GeoJSONPolygonGenericSchema = (positionSchema) => z
    .looseObject({
    ...GeoJSONGeometryBaseSchema(positionSchema).shape,
    type: z.literal(GeoJSONGeometryTypeSchema.enum.Polygon),
    // We allow an empty coordinates array
    // > GeoJSON processors MAY interpret Geometry objects with empty "coordinates"
    //   arrays as null objects. (RFC 7946, section 3.1)
    // > A linear ring is a closed LineString with four or more positions (RFC 7946, section 3.1.6)
    coordinates: z.array(positionSchema.array().min(4)),
})
    .check((ctx) => {
    // Skip remaining checks if coordinates array is empty
    if (!ctx.value.coordinates.length) {
        return;
    }
    if (!validDimensionsForPositionGrid(ctx.value)) {
        ctx.issues.push(getInvalidDimensionIssue(ctx));
        return;
    }
    if (!validPolygonRings(ctx.value)) {
        ctx.issues.push(getInvalidPolygonLinearRingIssue(ctx));
        return;
    }
    if (!validBBoxForPositionGrid(ctx.value)) {
        ctx.issues.push(getInvalidBBoxIssue(ctx));
        return;
    }
});
export const GeoJSONPolygonSchema = GeoJSONPolygonGenericSchema(GeoJSONPositionSchema);
export const GeoJSON2DPolygonSchema = GeoJSONPolygonGenericSchema(GeoJSON2DPositionSchema);
export const GeoJSON3DPolygonSchema = GeoJSONPolygonGenericSchema(GeoJSON3DPositionSchema);
