// zod-geojson/lib/geometry/validation/dimension.mjs
export const getInvalidDimensionIssue = (ctx) => ({
    code: "custom",
    message: "Invalid dimensions. All positions in the geometry must have the same dimension.",
    input: ctx.value,
    continue: true,
});
export const getInvalidGeometryCollectionDimensionIssue = (ctx) => ({
    code: "custom",
    message: "Invalid geometry collection dimensions. All geometries must have the same dimension.",
    input: ctx.value,
    continue: true,
});
export function validDimensionsForPositionList({ coordinates }) {
    const dimension = coordinates[0].length;
    return sameDimensionsForPositions(dimension)(coordinates);
}
export function validDimensionsForPositionGrid({ coordinates }) {
    let dimension = coordinates[0][0].length;
    return sameDimensionsForPositionGrid(dimension)(coordinates);
}
export function validDimensionsForPositionGridList({ coordinates }) {
    let dimension = coordinates[0][0][0].length;
    return sameDimensionsForPositionGrids(dimension)(coordinates);
}
export function validDimensionsForCollection({ geometries }) {
    if (geometries == null)
        return false;
    let dimension = getDimensionForGeometry(geometries[0]);
    return geometries.slice(1).every((geometry) => getDimensionForGeometry(geometry) === dimension);
}
export function getDimensionForGeometry(geometry) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    switch (geometry.type) {
        case "Point":
            return (_b = (_a = geometry.coordinates) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0;
        case "MultiPoint":
        case "LineString":
            return (_d = (_c = geometry.coordinates) === null || _c === void 0 ? void 0 : _c[0].length) !== null && _d !== void 0 ? _d : 0;
        case "MultiLineString":
        case "Polygon":
            return (_f = (_e = geometry.coordinates) === null || _e === void 0 ? void 0 : _e[0][0].length) !== null && _f !== void 0 ? _f : 0;
        case "MultiPolygon":
            return (_h = (_g = geometry.coordinates) === null || _g === void 0 ? void 0 : _g[0][0][0].length) !== null && _h !== void 0 ? _h : 0;
        case "GeometryCollection":
            return getDimensionForGeometry(geometry.geometries[0]);
    }
}
function sameDimensionsForPosition(dimension) {
    return (position) => position.length === dimension;
}
function sameDimensionsForPositions(dimension) {
    return (positions) => positions.every(sameDimensionsForPosition(dimension));
}
function sameDimensionsForPositionGrid(dimension) {
    return (positionGrid) => positionGrid.every(sameDimensionsForPositions(dimension));
}
function sameDimensionsForPositionGrids(dimension) {
    return (positionGrids) => positionGrids.every(sameDimensionsForPositionGrid(dimension));
}
