// zod-geojson/lib/geometry/validation/linear_ring.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInvalidMultiPolygonLinearRingIssue = exports.getInvalidPolygonLinearRingIssue = void 0;
exports.validPolygonRings = validPolygonRings;
exports.validMultiPolygonLinearRings = validMultiPolygonLinearRings;
const getInvalidPolygonLinearRingIssue = (ctx) => ({
    code: "custom",
    message: "Invalid polygon. Each ring inside a polygon must form a linear ring.",
    input: ctx.value,
    continue: true,
});
exports.getInvalidPolygonLinearRingIssue = getInvalidPolygonLinearRingIssue;
const getInvalidMultiPolygonLinearRingIssue = (ctx) => ({
    code: "custom",
    message: "Invalid multi polygon. Each polygon inside the multi polygon must be made out of linear rings.",
    input: ctx.value,
    continue: true,
});
exports.getInvalidMultiPolygonLinearRingIssue = getInvalidMultiPolygonLinearRingIssue;
function validLinearRing(linearRing) {
    const firstPosition = linearRing[0];
    const lastPosition = linearRing[linearRing.length - 1];
    return firstPosition.every((value, index) => value === lastPosition[index]);
}
function validPolygonRings({ coordinates: rings }) {
    return rings.every(validLinearRing);
}
function validMultiPolygonLinearRings({ coordinates }) {
    return coordinates.every((polygon) => validPolygonRings({ coordinates: polygon }));
}
