// zod-geojson/lib/feature.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeoJSON3DFeatureSchema = exports.GeoJSON2DFeatureSchema = exports.GeoJSONFeatureSchema = exports.GeoJSONFeatureGenericSchema = void 0;
const z = __importStar(require("zod"));
const base_1 = require("./base.js");
const geometry_1 = require("./geometry/geometry.js");
const position_1 = require("./geometry/position.js");
const bbox_1 = require("./geometry/validation/bbox.js");
const properties_1 = require("./properties.js");
const type_1 = require("./type.js");
const bbox_2 = require("./validation/bbox.js");
const GeoJSONFeatureGenericSchema = (positionSchema, propertiesSchema, geometrySchema) => z
    .looseObject({
    ...(0, base_1.GeoJSONBaseSchema)(positionSchema).shape,
    id: z.string().or(z.number()).optional(),
    type: z.literal(type_1.GeoJSONTypeSchema.enum.Feature),
    geometry: geometrySchema,
    properties: propertiesSchema,
    coordinates: z.never({ error: "GeoJSON Feature cannot have a 'coordinates' key" }).optional(),
    features: z.never({ error: "GeoJSON Feature cannot have a 'features' key" }).optional(),
    geometries: z.never({ error: "GeoJSON Feature cannot have a 'geometries' key" }).optional(),
})
    .check((ctx) => {
    if (!(0, bbox_2.validBBoxForFeature)(ctx.value)) {
        ctx.issues.push((0, bbox_1.getInvalidBBoxIssue)(ctx));
    }
});
exports.GeoJSONFeatureGenericSchema = GeoJSONFeatureGenericSchema;
exports.GeoJSONFeatureSchema = (0, exports.GeoJSONFeatureGenericSchema)(position_1.GeoJSONPositionSchema, properties_1.GeoJSONPropertiesSchema.nullable(), geometry_1.GeoJSONGeometrySchema);
exports.GeoJSON2DFeatureSchema = (0, exports.GeoJSONFeatureGenericSchema)(position_1.GeoJSON2DPositionSchema, properties_1.GeoJSONPropertiesSchema.nullable(), geometry_1.GeoJSON2DGeometrySchema);
exports.GeoJSON3DFeatureSchema = (0, exports.GeoJSONFeatureGenericSchema)(position_1.GeoJSON3DPositionSchema, properties_1.GeoJSONPropertiesSchema.nullable(), geometry_1.GeoJSON3DGeometrySchema);
