// zod-geojson/lib/feature_collection.mjs
import * as z from "zod";
import { GeoJSONBaseSchema } from "./base.mjs";
import { GeoJSONFeatureGenericSchema } from "./feature.mjs";
import { GeoJSON2DPositionSchema, GeoJSON3DPositionSchema, GeoJSONPositionSchema, } from "./geometry/position.mjs";
import { getInvalidBBoxIssue } from "./geometry/validation/bbox.mjs";
import { GeoJSONPropertiesSchema } from "./properties.mjs";
import { GeoJSONTypeSchema } from "./type.mjs";
import { validBBoxForFeatureCollection } from "./validation/bbox.mjs";
import { getInvalidFeatureCollectionDimensionsIssue, validDimensionsForFeatureCollection, } from "./validation/dimension.mjs";
import { GeoJSON2DGeometrySchema, GeoJSON3DGeometrySchema, GeoJSONGeometrySchema, } from "./geometry/geometry.mjs";
export const GeoJSONFeatureCollectionGenericSchema = (positionSchema, propertiesSchema, geometrySchema) => z
    .looseObject({
    ...GeoJSONBaseSchema(positionSchema).shape,
    type: z.literal(GeoJSONTypeSchema.enum.FeatureCollection),
    features: z.array(GeoJSONFeatureGenericSchema(positionSchema, propertiesSchema, geometrySchema)),
    coordinates: z.never({ error: "GeoJSON feature collection cannot have a 'coordinates' key" }).optional(),
    geometry: z.never({ error: "GeoJSON feature collection cannot have a 'geometry' key" }).optional(),
    properties: z.never({ error: "GeoJSON feature collection cannot have a 'properties' key" }).optional(),
    geometries: z.never({ error: "GeoJSON feature collection cannot have a 'geometries' key" }).optional(),
})
    .check((ctx) => {
    if (!ctx.value.features.length) {
        return;
    }
    if (!validDimensionsForFeatureCollection(ctx.value)) {
        ctx.issues.push(getInvalidFeatureCollectionDimensionsIssue(ctx));
        return;
    }
    if (!validBBoxForFeatureCollection(ctx.value)) {
        ctx.issues.push(getInvalidBBoxIssue(ctx));
        return;
    }
});
export const GeoJSONFeatureCollectionSchema = GeoJSONFeatureCollectionGenericSchema(GeoJSONPositionSchema, GeoJSONPropertiesSchema.nullable(), GeoJSONGeometrySchema);
export const GeoJSON2DFeatureCollectionSchema = GeoJSONFeatureCollectionGenericSchema(GeoJSON2DPositionSchema, GeoJSONPropertiesSchema.nullable(), GeoJSON2DGeometrySchema);
export const GeoJSON3DFeatureCollectionSchema = GeoJSONFeatureCollectionGenericSchema(GeoJSON3DPositionSchema, GeoJSONPropertiesSchema.nullable(), GeoJSON3DGeometrySchema);
