// zod-geojson/lib/feature.mjs
import * as z from "zod";
import { GeoJSONBaseSchema } from "./base.mjs";
import { GeoJSON2DGeometrySchema, GeoJSON3DGeometrySchema, GeoJSONGeometrySchema, } from "./geometry/geometry.mjs";
import { GeoJSON2DPositionSchema, GeoJSON3DPositionSchema, GeoJSONPositionSchema, } from "./geometry/position.mjs";
import { getInvalidBBoxIssue } from "./geometry/validation/bbox.mjs";
import { GeoJSONPropertiesSchema } from "./properties.mjs";
import { GeoJSONTypeSchema } from "./type.mjs";
import { validBBoxForFeature } from "./validation/bbox.mjs";
export const GeoJSONFeatureGenericSchema = (positionSchema, propertiesSchema, geometrySchema) => z
    .looseObject({
    ...GeoJSONBaseSchema(positionSchema).shape,
    id: z.string().or(z.number()).optional(),
    type: z.literal(GeoJSONTypeSchema.enum.Feature),
    geometry: geometrySchema,
    properties: propertiesSchema,
    coordinates: z.never({ error: "GeoJSON Feature cannot have a 'coordinates' key" }).optional(),
    features: z.never({ error: "GeoJSON Feature cannot have a 'features' key" }).optional(),
    geometries: z.never({ error: "GeoJSON Feature cannot have a 'geometries' key" }).optional(),
})
    .check((ctx) => {
    if (!validBBoxForFeature(ctx.value)) {
        ctx.issues.push(getInvalidBBoxIssue(ctx));
    }
});
export const GeoJSONFeatureSchema = GeoJSONFeatureGenericSchema(GeoJSONPositionSchema, GeoJSONPropertiesSchema.nullable(), GeoJSONGeometrySchema);
export const GeoJSON2DFeatureSchema = GeoJSONFeatureGenericSchema(GeoJSON2DPositionSchema, GeoJSONPropertiesSchema.nullable(), GeoJSON2DGeometrySchema);
export const GeoJSON3DFeatureSchema = GeoJSONFeatureGenericSchema(GeoJSON3DPositionSchema, GeoJSONPropertiesSchema.nullable(), GeoJSON3DGeometrySchema);
