// zod-geojson/lib/feature_collection.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeoJSON3DFeatureCollectionSchema = exports.GeoJSON2DFeatureCollectionSchema = exports.GeoJSONFeatureCollectionSchema = exports.GeoJSONFeatureCollectionGenericSchema = void 0;
const z = __importStar(require("zod"));
const base_1 = require("./base.js");
const feature_1 = require("./feature.js");
const position_1 = require("./geometry/position.js");
const bbox_1 = require("./geometry/validation/bbox.js");
const properties_1 = require("./properties.js");
const type_1 = require("./type.js");
const bbox_2 = require("./validation/bbox.js");
const dimension_1 = require("./validation/dimension.js");
const geometry_1 = require("./geometry/geometry.js");
const GeoJSONFeatureCollectionGenericSchema = (positionSchema, propertiesSchema, geometrySchema) => z
    .looseObject({
    ...(0, base_1.GeoJSONBaseSchema)(positionSchema).shape,
    type: z.literal(type_1.GeoJSONTypeSchema.enum.FeatureCollection),
    features: z.array((0, feature_1.GeoJSONFeatureGenericSchema)(positionSchema, propertiesSchema, geometrySchema)),
    coordinates: z.never({ error: "GeoJSON feature collection cannot have a 'coordinates' key" }).optional(),
    geometry: z.never({ error: "GeoJSON feature collection cannot have a 'geometry' key" }).optional(),
    properties: z.never({ error: "GeoJSON feature collection cannot have a 'properties' key" }).optional(),
    geometries: z.never({ error: "GeoJSON feature collection cannot have a 'geometries' key" }).optional(),
})
    .check((ctx) => {
    if (!ctx.value.features.length) {
        return;
    }
    if (!(0, dimension_1.validDimensionsForFeatureCollection)(ctx.value)) {
        ctx.issues.push((0, dimension_1.getInvalidFeatureCollectionDimensionsIssue)(ctx));
        return;
    }
    if (!(0, bbox_2.validBBoxForFeatureCollection)(ctx.value)) {
        ctx.issues.push((0, bbox_1.getInvalidBBoxIssue)(ctx));
        return;
    }
});
exports.GeoJSONFeatureCollectionGenericSchema = GeoJSONFeatureCollectionGenericSchema;
exports.GeoJSONFeatureCollectionSchema = (0, exports.GeoJSONFeatureCollectionGenericSchema)(position_1.GeoJSONPositionSchema, properties_1.GeoJSONPropertiesSchema.nullable(), geometry_1.GeoJSONGeometrySchema);
exports.GeoJSON2DFeatureCollectionSchema = (0, exports.GeoJSONFeatureCollectionGenericSchema)(position_1.GeoJSON2DPositionSchema, properties_1.GeoJSONPropertiesSchema.nullable(), geometry_1.GeoJSON2DGeometrySchema);
exports.GeoJSON3DFeatureCollectionSchema = (0, exports.GeoJSONFeatureCollectionGenericSchema)(position_1.GeoJSON3DPositionSchema, properties_1.GeoJSONPropertiesSchema.nullable(), geometry_1.GeoJSON3DGeometrySchema);
