// zod-geojson/lib/validation/dimension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInvalidFeatureCollectionDimensionsIssue = void 0;
exports.validDimensionsForFeatureCollection = validDimensionsForFeatureCollection;
const dimension_1 = require("../geometry/validation/dimension.js");
const types_1 = require("./types.js");
const getInvalidFeatureCollectionDimensionsIssue = (ctx) => ({
    code: "custom",
    message: "Invalid dimensions. All features in feature collection must have the same dimension.",
    input: ctx.value,
    continue: true,
});
exports.getInvalidFeatureCollectionDimensionsIssue = getInvalidFeatureCollectionDimensionsIssue;
function validDimensionsForFeatureCollection(collection) {
    const geometries = (0, types_1.getGeometries)(collection);
    if (geometries.length === 0)
        return true;
    const dimension = (0, dimension_1.getDimensionForGeometry)(geometries[0]);
    return geometries.slice(1).every((geometry) => (0, dimension_1.getDimensionForGeometry)(geometry) === dimension);
}
